### PR TITLE
Finish chapter 6: modelling user.

### PR DESCRIPTION
**Pull request by Naodab**
## Done chapter 6 Modeling User
**Topic**:
- User Model (database migration, CRUD).
- User validations (presence, length, format, uniqueness).
- Adding a secure password (hashed password, has_secure_password_method, ,minimum password).

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+class User < ApplicationRecord
+  before_save :downcase_email
+
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+
+  validates :name, presence: true, length: { maximum: 50 }
```
[the remark this answers is not in the document]
Comment (Naodab) [reply app/models/user.rb:6]: dạ thêm một call back ở before save
```Ruby
before_save :upcase_name

private

def upcase_name
     self.name = name.upcase
end
```

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  validates :email,
+            presence: true,
+            length: { maximum: 255 },
+            format: { with: VALID_EMAIL_REGEX },
+            uniqueness: true
+  validates :password,
```
[the remark this answers is not in the document]
Comment (Naodab) [reply app/models/user.rb:12]: Dạ  anh

File: test/models/user_test.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  def setup
+    @user = User.new(name: "Example User", email: "user@example.com",
+                      password: "foobar2", password_confirmation: "foobar2")
+  end
+
+  test "valid user" do
+    assert @user.valid?
+  end
+
+  test "invalid without name" do
```
Comment (tamnt39) [inline test/models/user_test.rb:13]: Mình gộp lại dạng này cho đỡ phải viết nhiều text:
```
test "invalid email format edge cases" do
  invalid_emails = [
    "user@.com",
    "@example.com",
    "user@example.",
    "user@-example.com"
  ]
  invalid_emails.each do |email|
    @user.email = email
    assert_not @user.valid?, "#{email} should be invalid"
  end
end
```

Comment (Naodab) [reply test/models/user_test.rb:13]: Dạ anh, để em chỉnh sửa lại

**Review by tamnt39 — APPROVED** [review]
LGTM!
